STM32: Add Ethernet PHY mode and speed macros and update DTS requirements

marwaiehm-st opened the pull request:
- Added required property for phy-connection-type in the DTS .
- Updated the DTS file to include phy-connection-type with a default value of "rmii".
- Added MAC_NODE definition for easier reference to the MAC node.
- Simplified the STM32_ETH_PHY_MODE macro by removing redundant checks for RMII.
- The macro now defaults to ETH_MEDIA_INTERFACE_RMII if the phy_connection_type is not MII.
- Simplified the STM32_ETH_SPEED macro to always return ETH_SPEED_100M.
- Updated the eth_initialize function to use the STM32_ETH_SPEED macro for setting the speed.
- Updated the eth0_data initialization to use the STM32_ETH_PHY_MODE macro for setting the media interface.
- Updated migration guide 4.2 for STM32 HAL Ethernet changes


This change improves code readability and maintainability by reducing redundancy and simplifying the logic. It also ensures that the phy-connection-type is explicitly defined in the DTS file.

marwaiehm-st committed:
drivers: ethernet: remove ETH_STM32_HAL_MII

Removed `ETH_STM32_HAL_MII` configuration option
from `Kconfig.stm32_hal`.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
boards: Specify PHY connection type

This commit updates the device tree source files
for all STM32 boards to specify the PHY connection
type for the Ethernet configuration.
Added the `phy-connection-type` property
with the value "rmii" to the Ethernet node.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
drivers: ethernet: Update Ethernet PHY mode and speed configuration

- Added macros `STM32_ETH_PHY_MODE` to determine
  the PHY mode and speed from the `phy_connection_type`
  property in the device tree.
- Updated the `eth0_data` structure to use `STM32_ETH_PHY_MODE`
  for setting the media interface.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
dts: bindings: ethernet: Add required phy-connection-type

This commit updates the STM32 Ethernet binding to include
the `phy-connection-type` property as required.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -88,6 +88,14 @@ static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? \
+		ETH_MEDIA_INTERFACE_MII : ETH_MEDIA_INTERFACE_RMII)
+
+#define STM32_ETH_SPEED(node_id) ETH_SPEED_100M
```
this define is not needed.

maass-hamburg commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -70,11 +70,6 @@ config ETH_STM32_HAL_PHY_ADDRESS
 	help
 	  The phy address to use.
 
-config ETH_STM32_HAL_MII
-	bool "Use MII interface"
-	help
-	  Use the MII physical interface instead of RMII.
-
```
mention in migration guide

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1321,7 +1329,7 @@ static struct eth_stm32_hal_dev_data eth0_data = {
 #else
```
should then be a ``BUILD_ASSERT()``

marwaiehm-st replied:
Done

marwaiehm-st committed:
doc: update migration guide 4.2 for STM32 HAL Ethernet changes

Removed the Kconfig option `ETH_STM32_HAL_MII`.
PHY interface type should now be set using
the `phy-connection-type` property in the device tree.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st commented on boards/st/stm32h745i_disco/stm32h745i_disco_stm32h745xx_m7.dts:
```diff
@@ -144,6 +144,7 @@
 		     &eth_txd0_pg13
 		     &eth_rx_er_pi10>;
 	pinctrl-names = "default";
+	phy-connection-type = "mii";
```
Done

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1311,6 +1317,10 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
 };
 
+BUILD_ASSERT(DT_ENUM_HAS_VALUE(MAC_NODE, phy_connection_type, mii) || \
```
``\`` at the end of line not needed

maass-hamburg commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -46,3 +46,13 @@ Modules
 
 Architectures
 *************
+
+Ethernet
+********
+
+* STM32 HAL Ethernet Changes:
+
+  * Removed the Kconfig option ``ETH_STM32_HAL_MII``:
+    * This option was used to select the MII physical interface instead of RMII.
+    * The selection of the PHY interface type should now be handled through
+      the ``phy-connection-type`` property in the device tree.
```
this belongs to ``Device Drivers and Devicetree``, take a look at ``migration-guide-4.1.rst`` for reference.

can also be a bit shorter

mathieuchopstm reviewed (commented):
Minor nit: commit `drivers: ethernet: remove ETH_STM32_HAL_MII` should come after commit `drivers: ethernet: Update Ethernet PHY mode and speed configuration`.

Otherwise, LGTM.

maass-hamburg reviewed (approved):
LGTM
